Feature/message wrapper

bfc5288 opened the pull request:
## Description

This PR implements https://github.com/scimma/architecture/issues/31, namely to restructure messages so that they're wrapped as:
```
{
    "format ": "message_format"
    "content": " message_content"
}
```

This allows `hop subscribe` to more reliably classify incoming messages without dubious substring parsing (as in #68). This relies on the publisher to specify their message format during the send; otherwise the message is sent as an unstructured blob which `subscribe` handles in turn (hence implementing #75).

Currently, `hop publish` supports publishing multiple messages if they're of the same format via:
```
hop publish URL -f <file-format> file1 file2 file3 ....
```
An alternative would be to allow the publisher to send multiple messages of different formats, e.g.:
```
hop publish URL -f format1 file1 -f format2 file2 ....
```
which could be implemented with a few changes to the publish CLI.

To prepare to generalize to future message formats, I also renamed the `gcn` format to `circular` and ensured that `gcn` wasn't used to refer to general message variables (`message` is used instead).

I added unit tests for the new functionality and modified existing tests when appropriate, though there are a few changes we could still consider (e.g., parametrizing tests in `test_cli` to reduce repeated code [see test_subscribe]; potentially moving more towards the `expected_data`/`test_data` structure instead of using definitions in `conftest`, etc.).

## Checklist

* [x] All new functions and classes are documented and adhere to Google doc style (3.8.3-3.8.6 of [this](http://google.github.io/styleguide/pyguide.html#383-functions-and-methods) document)
* [x] Add/update sphinx documentation with any relevant changes.
* [x] Add/update pytest-style tests in `/tests`, ensuring sufficient code coverage.
* [x] `make test` runs without errors.
* [x] `make lint` doesn't give any warnings.
* [x] `make format` doesn't give any code formatting suggestions.
* [x] `make doc` runs without errors and generated docs render correctly.
* [x] Check that CI pipeline run on this PR passes all stages.
* [ ] Review signoff by at least one developer.

NOTE: If this PR relates to a release, open and reference an issue with the Release checklist template.

spenczar commented on hop/publish.py:
```diff
@@ -19,37 +20,54 @@
+        for message_file in args.message:
+
+            if message_format == "circular":
+                with open(message_file, "r") as f:
+                    message_model = GCNCircular.from_email(f)
+            elif message_format == "voevent":
+                with open(message_file, "rb") as f:
+                    message_model = VOEvent.from_xml(f)
+            elif message_format == "blob":
+                with open(message_file, "r") as f:
+                    message_model = message_blob.from_file(f)
             else:
-                raise ValueError(f"File extension {extension} not recognized")
+                warnings.warn("Warning: format not recognized. Sending as unstructured blob")
+                with open(message_file, "r") as f:
+                    message_model = message_blob.from_file(f)
```
One way to get around the `"rb"` vs `"r"` file mode differences is for each function in the `model_loader` map to take a filename, so like 
```python
formats = {
    "circular": GCNCircular.from_email_file,
    "voevent": VOEvent.from_xml_file,
    "blob": message_blob.from_file,
}
...
if message_format in formats:
    loader = formats[message_format]
else:
    # raise the warning, etc
for message_file in args.message:
    message_model = loader(message_file)
```
```python
class VOEvent:
   ...
    @classmethod
    def from_xml_file(cls, filename):
        with open(filename, "rb") as f:
            return cls.from_xml(f)
```

bfc5288 replied:
Thanks for the pointers. I implemented the mapping with @spenczar's structure so that the file loading can now happen from `models.py`. (I also used the mapping structure in a similar setup to generalize the unit tests)

spenczar commented on hop/subscribe.py:
```diff
@@ -1,64 +1,64 @@
+    try:
+        fmt = message["format"]
+        content = message["content"]
+    except TypeError:
```
I think also want to check for `KeyError`, in case `message` is a dict but it just doesn't have those keys.

bfc5288 replied:
Good catch, I added this and a test for both errors.

bfc5288 commented:
> Looks good, @bfc5288! I added some review comments/suggestions.
Thanks for the comments! I added revisions.

For unit test restructuring, what seemed most general was to parametrize each test with a list of message formats. Then, the test accesses a nested dictionary fixture from `conftest`, which holds the parameters for each message type (name, model, file, and text).

This gets around being unable to have fixtures in parametrizations (e.g., parametrizing test_cli_publish with the `voevent_text` fixture), and allows `test_cli` and `test_subscribe` to be easily extended for future message types (i.e., putting message name in the parametrization, and its parameters in `conftest`). For now, `test_models` probably still has to have separate test functions for each message type.

I'm wrapping up the changes and will update the docs next.

myNameIsPatrick commented on hop/models.py:
```diff
@@ -103,12 +139,79 @@ def from_email(cls, email_input):
         return cls(
-            header={title.lower(): content for title, content in msg.items()},
-            body=msg.get_payload(),
+            header={title.lower(): content for title,
+                    content in message.items()},
```
Is there any way to get this to fit into one line instead? It's harder to see that it's a dict comprehension when it's split like this.

bfc5288 replied:
Yep, good catch. That was an auto-format change which definitely makes it less readable. Fixed.

myNameIsPatrick commented on hop/subscribe.py:
```diff
@@ -1,64 +1,67 @@
+        raise KeyError("Message does not contain format/content keys")
+
+    # create the dataclass model appropriate for the message format
+    model_creator = {"circular": GCNCircular,
+                     "voevent": VOEvent,
+                     "blob": MessageBlob,
+                     }
```
Same formatting as above if applicable.

myNameIsPatrick commented on hop/publish.py:
```diff
@@ -38,18 +48,19 @@ def _main(args=None):
-            # publish GCN
-            s.write(gcn.asdict())
+
+        model_loader = {"circular": GCNCircular.from_email_file,
+                        "voevent": VOEvent.from_xml_file,
+                        "blob": MessageBlob.from_text,
+                        }
```
I think it may be cleaner if we move this outside of the stream context. Also, I think it may be nicer if it's formatted as such:
```
model_loader = {
    "circular": GCNCircular.from_email_file,
    "voevent": VOEvent.from_xml_file,
    "blob": MessageBlob.from_text,
}
```

PEP8 is a little more loose on this requirement but there's already precedent for this formatting in the argparse arguments and I'd like to be consistent.

myNameIsPatrick commented on tests/conftest.py:
```diff
@@ -193,3 +196,43 @@ def voevent_fileobj():
+@pytest.fixture(scope="session")
+def message_parameters():
+    # generalize model_name, expected_model, and test_file
+    return {"circular": ["GCNCircular", models.GCNCircular, "example_gcn.gcn3", GCN_CIRCULAR],
+            "voevent": ["VOEvent", models.VOEvent, "example_voevent.xml", VOEVENT_XML],
+            "blob": ["MessageBlob", models.MessageBlob, "example_blob.txt", MESSAGE_BLOB],
+            }
+
+
+@pytest.fixture(scope="session")
+def message_parameters_dict():
+    # generalize model_name, expected_model, and test_file
+    return {"circular": {"model_name": "GCNCircular",
+                         "expected_model": models.GCNCircular,
+                         "test_file": "example_gcn.gcn3",
+                         "model_text": GCN_CIRCULAR,
+                         },
+            "voevent": {"model_name": "VOEvent",
+                        "expected_model": models.VOEvent,
+                        "test_file": "example_voevent.xml",
+                        "model_text": VOEVENT_XML.encode(),
+                        },
+            "blob":     {"model_name": "MessageBlob",
+                         "expected_model": models.MessageBlob,
+                         "test_file": "example_blob.txt",
+                         "model_text": MESSAGE_BLOB,
+                         },
+            }
```
Same formatting as above.
